### PR TITLE
Support Cardano key derivation according to CIP3-Icarus

### PR DESCRIPTION
**Pull request by PeterBenc**
### What is the current state of things and why does it need to change?

- currently, this lib supports `secp256k1` and `ed25519`
- while developing a Snap for Cardano, we realized that keys derived by snap, using (`getBip32Entropy`) are different from keys that would standard Cardano wallet give, for the same mnemonic
- this inconsistency could lead to confusion and interoperability issues
- currently developed Cardano snap, relies on these changes, and is blocked by this. (note that changes need to be made also to `metamask-snaps` and `metamask-extension` repos, which are ready, awaiting feedback on this)

### What is the solution your changes offer and how does it work?

- by supporting derivation which is consistent with Cardano standard, we enable current Cardano users to use their existing mnemonics in Metamask and get the same wallet they used until now.
- similarily, we would enable current Metamask users to use their mnemonic in standard Cardano wallets and retrieve same addresses as they would through Cardano-Metamask snap.

For this reasons
- we add a new derivation `curve`, called `ed25519Bip32`, this curve implements a slight modification of how standard `ed25519` is used in SLIP10. 

- we add a new `deriver`, called `cip3Icarus` which derives keys according to `CIP3` (which is the standard for Cardano)

- from the interface point of view, these changes allow us to call 

  ```js
  await deriveKeyFromPath({
     path: ["cip3Icarus:1852'", "cip3Icarus: 1815'"],
     curve: 'ed25519Bip32',
  })
  ```

  which returns keys consistent with Cardano derivation standard

**A longer explanation, also present in the code**

[CIP-3](https://github.com/cardano-foundation/CIPs/blob/09d7d8ee1bd64f7e6b20b5a6cae088039dce00cb/CIP-0003/CIP-0003.md) defines standards for deriving keys on Cardano.
 
 Key attributes

   * Root/Master key is derived from entropy, not seed. For this implementation we work with **Icarus** standard as it is the most widely used. See https://github.com/cardano-foundation/CIPs/blob/09d7d8ee1bd64f7e6b20b5a6cae088039dce00cb/CIP-0003/Icarus.md.
     
   * HD node consists of a 64 byte private key, 32 byte public key and 32 byte chain code. See https://github.com/cardano-foundation/CIPs/blob/09d7d8ee1bd64f7e6b20b5a6cae088039dce00cb/CIP-0003/CIP-0003.md#master-key-generation.
     
   * For derivation of BIP32 HD nodes, it uses modified implementation of ed25519 called BIP32-Ed25519. See https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf.

### Are there any issues or other links reviewers should consult to understand this pull request better? For instance:

All sources are referenced in the code, and explained quite extensively, namely

- [CIP3](https://github.com/cardano-foundation/CIPs/blob/09d7d8ee1bd64f7e6b20b5a6cae088039dce00cb/CIP-0003/CIP-0003.md)
- [ed25519Bip32](https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf)

Reference implementations

- [cardano-js-sdk](https://github.com/input-output-hk/cardano-js-sdk/tree/master/packages/crypto)
- [cardano-crypto.js](https://github.com/vacuumlabs/cardano-crypto.js)

### CODE REVIEW

File: src/utils.ts
Comment (PeterBenc) [inline src/utils.ts:444]: Needed later, to encode keyindex, for Cardano, the values are `littleEndian`

**Review by PeterBenc — COMMENTED** [review]
Please review by commits, thank you :pray:

File: src/derivers/cip3Icarus.ts
Comment (Mrtenz) [inline src/derivers/cip3Icarus.ts:1]: Sorry for the delay on this PR!

I'm not familiar with CIP-3 and the different key generation methods it describes. Is Icarus the main standard used by Cardano? If so, what do you think about renaming this to `cip3` instead of `cip3Icarus`? I think that simplifies the public API a little bit (`["cip3:0'", "cip3:..."]` versus `["cip3Icarus:0'", "cip3Icarus:..."]`). Icarus seems to be the recommended method by CIP-3, and if other methods need to be added in the future, we could add a suffix for those (like `cip3Trezor` etc.).

File: src/utils.ts
Comment (Mrtenz) [reply src/utils.ts:382]: I would prefer adding a new parameter, so we can keep the validation as strict as possible.

Comment (PeterBenc) [reply src/utils.ts:382]: Ok, I ll do that :+1:

File: src/derivers/cip3Icarus.ts
Comment (PeterBenc) [reply src/derivers/cip3Icarus.ts:1]: Yes, `Icarus` is the main standard Cardano uses, but [Ledger](https://github.com/cardano-foundation/CIPs/blob/09d7d8ee1bd64f7e6b20b5a6cae088039dce00cb/CIP-0003/Ledger.md) also uses another significant one.
We were discussing this internally too, and initially, this was as you are proposing, just `cip3`. We then decided to be more specific and use `cip3Icarus`. When working on this feature, I found that the information is often not as straightforward as it should be, and bringing more clarity certainly does not hurt, even at the expense of the bit longer name.

Anyway, let me know what you think, I don't have a strong opinion about this, so we can change it to just `cip3` if you insist :+1:

Comment (Mrtenz) [reply src/derivers/cip3Icarus.ts:1]: I feel like simplifying it in the API and providing good docs around this is better in this case.

Comment (PeterBenc) [reply src/derivers/cip3Icarus.ts:1]: Ok, sure, will do that.

**Comment by Mrtenz** [comment]
At first glance this looks great. I will be doing a proper review next week, thanks for your patience!

File: src/curves/ed25519Bip32.ts
Comment (Mrtenz) [inline src/curves/ed25519Bip32.ts:18]: Should we call this "bip32-ed25519" (since that's what the spec is called)?

File: src/curves/ed25519Bip32.ts
Comment (Mrtenz) [inline src/curves/ed25519Bip32.ts:1]: Can you add JSDoc to all functions here?

File: src/derivers/bip39.ts
Comment (Mrtenz) [inline src/derivers/bip39.ts:119]:
```suggestion
  rootNode[31] |= 0b0100_0000;
  /* eslint-enable no-bitwise */
```

File: src/derivers/cip3Icarus.test.ts
Comment (Mrtenz) [inline src/derivers/cip3Icarus.test.ts]:
```suggestion
      });

```

File: src/derivers/cip3Icarus.test.ts
Comment (Mrtenz) [inline src/derivers/cip3Icarus.test.ts]:
```suggestion
      });

```

File: src/derivers/cip3Icarus.test.ts
Comment (Mrtenz) [inline src/derivers/cip3Icarus.test.ts]:
```suggestion
      });

```

File: src/derivers/cip3Icarus.test.ts
Comment (Mrtenz) [inline src/derivers/cip3Icarus.test.ts]:
```suggestion
      });

```

File: src/derivers/cip3Icarus.test.ts
Comment (Mrtenz) [inline src/derivers/cip3Icarus.test.ts]:
```suggestion
      });

```

File: src/derivers/cip3Icarus.ts
Comment (Mrtenz) [inline src/derivers/cip3Icarus.ts:1]: Can you add JSDoc to the functions here?

**Review by Mrtenz — CHANGES_REQUESTED** [review]
I didn't get through everything yet, but here's some initial comments for you. I'm not familiar with CIP-3, so I will take some time to read more about it to properly review the algorithms added here.

Some nits related to the code style we use. Sorry for that!

File: test/vectors/derivation.json
Comment (PeterBenc) [inline test/vectors/derivation.json:1]: Did this in separate commit so I can squash it easier

File: src/derivers/bip39.ts
Comment (PeterBenc) [inline src/derivers/bip39.ts:38]: You haven't asked for this, but, I noticed it now, the switch is IMO safer

**Review by PeterBenc — COMMENTED** [review]
@Mrtenz I was initially waiting for your full review, but then thought that you might be waiting for the `cip3Icarus` => `cip3` refactor, so I did it. Your review came literally one second after I pushed :D. Sorry for that, please check these 3 commits too.

File: src/utils.ts
Comment (PeterBenc) [reply src/utils.ts]: https://github.com/MetaMask/key-tree/pull/158/commits/4f546810aa59355c6692f1f2ef1b6802aa141e37

File: src/derivers/cip3Icarus.ts
Comment (PeterBenc) [reply src/derivers/cip3Icarus.ts:1]: https://github.com/MetaMask/key-tree/pull/158/commits/a234957284d6637d99f27ffd014323c5e7e0c493

File: src/curves/ed25519Bip32.ts
Comment (PeterBenc) [reply src/curves/ed25519Bip32.ts]: I inlined `modN` and `modNLE` which IMO makes sense, it was too simple to have in separate functions. Thanks :+1: 
I left `bytesToNumberLE` and `multiplyWithBase` in separate functions. Even though they are used only once, it makes sense to me (especially for `multiplyWithBase`) to be separate, since they are more complex, and own function is IMO a good space for documenting what these cryptic transformations do in plain English. 

Let me know what you think, naturally, I don't insist, i can inline them if you do :+1:

File: src/curves/ed25519Bip32.ts
Comment (PeterBenc) [reply src/curves/ed25519Bip32.ts:18]: We had an internal discussion about this too. 

Regarding the `-`, we kept camel case, this is later imported with `import * as ed25519Bip32` which does not work with kebab case (afaik), and also is then put into object, which allows only camel case (we can disable eslint for that line ofc, but we decided that it's not worth it)

Initially, I called this `bip32Ed25519` but after discussions changed it to `ed25519Bip32` because 
> bip32 is "semantically" less important than the curve itself

Let me know what you think :)

File: src/derivers/cip3Icarus.ts
Comment (PeterBenc) [reply src/derivers/cip3Icarus.ts:1]: https://github.com/MetaMask/key-tree/pull/158/commits/f0479a603475c0b26f3389f614d77addd8c5a674

**Comment by PeterBenc** [comment]
@Mrtenz have you had time to look into `cip3`? 
PS: all relevant links are in the PR description :)

**Comment by Mrtenz** [comment]
> @Mrtenz have you had time to look into `cip3`? PS: all relevant links are in the PR description :)

Not yet, sorry about that. Will make time to look into this more this week!

File: src/curves/ed25519Bip32.ts
Comment (Mrtenz) [inline src/curves/ed25519Bip32.ts:86]:
```suggestion
  publicKey: Uint8Array,
  tweak: Uint8Array,
```

File: src/derivers/cip3.ts
Comment (Mrtenz) [inline src/derivers/cip3.ts:262]: Just for clarification: The spec describes a way to derive public keys for `i >= 2^31`, but we don't need to implement that because we can use private key derivation in that case?

File: src/derivers/cip3.ts
Comment (Mrtenz) [inline src/derivers/cip3.ts:354]:
```suggestion
    "Invalid parameters: Cannot derive hardened child keys without a private key.",
```

File: src/utils.ts
Comment (Mrtenz) [inline src/utils.ts:388]:
```suggestion
      `Invalid public key: The key must be a ${compressedPublicKeyLength}-byte, non-zero byte array.`,
```

**Review by Mrtenz — CHANGES_REQUESTED** [review]
I reviewed the cryptography to the best of my ability and it looks good. Just a few more minor nits. Once they are fixed I will approve!

**Comment by Mrtenz** [comment]
Can you rebase your branch?

**Comment by PeterBenc** [comment]
@Mrtenz the last couple of issues should be resolved. Should I squash the commits or do you want to do it yourself?

**Comment by Mrtenz** [comment]
`src/index.ts` is missing coverage(?), and you need to run `yarn lint:fix`. Looks good to me other than that!

**Comment by PeterBenc** [comment]
Should be fixed

Should I squash the commits into some reasonable chunks, would that be relevant for audit too? do you want to do it yourself? Let me know :)

**Comment by Mrtenz** [comment]
No need to. We want to do an audit for some other smaller changes since the last audit anyway, so I will just squash when merging this PR.

**Pull request merged**
